### PR TITLE
Frozen point high tilt calibration

### PR DESCRIPTION
**Pull request by d-beniwal**
# Summary

Adds midas_calibrate_v2.pipelines.frozen_point, a calibration path for detector geometries where the existing alternating pipeline (autocalibrate_pv) can converge cleanly to a self-consistent but wrong geometry — specifically large/only-roughly-known detector tilt combined with an off-detector beam centre.

Root problem: autocalibrate_pv extracts peak positions at the current trial geometry, refits, and re-extracts at the new geometry, repeated for a few outer iterations. If the starting geometry is far enough off, a biased extraction at a still-wrong geometry can compound across iterations into an "honest" wrong answer — the fit reports rc=0 and looks converged, but isn't. On the validating dataset (CeO2, Varex 2880×2880, tz≈14° tilt, beam centre off the active area), this is a real, reproducible failure mode.

# What's included

## Two new functions, both in midas_calibrate_v2/pipelines/frozen_point.py:

- autocalibrate_frozen_point — picks its point cloud once via a genuine local-intensity-maximum search directly on the raw image (forward/point_pick.py, no azimuthal binning, no windowed re-sampling at a moving trial geometry), freezes it, then does a single bounded least-squares fit. No re-extraction step exists for a bad early estimate to compound through. tx is force-frozen (unobservable from single-image ring radii — a genuine gauge freedom, not a numerical artifact). On the validating dataset, from the same rough seed that led autocalibrate_pv to the wrong basin, this recovers the correct geometry in ~3s.
- iterate_frozen_point_until_stable — for when there's no tilt guess at all. Re-seeds autocalibrate_frozen_point from its own result repeatedly, re-centering the fit bounds on the new estimate each time (this re-centering, not a fixed window, is what lets it walk many degrees from a bad start). Stops on a strict parameter-stability
  criterion — 5 consecutive iterations agreeing to 0.01° (tz/ty), 0.05px (BC), 5µm (Lsd) — deliberately with no
  strain gate. See the function's docstring for why: a loose tolerance was found to falsely accept a multi-iteration plateau that looked stable but was still ~10-12° from the true answer.

Validation (offline 13-point sweep over the starting tilt guess, documented in the function's docstring and in scratch_calibrate/large_tilt_calib/best_solution/README.md on this machine): converges to the identical, bit-reproducing fixed point from starting guesses spanning ~35° — including a completely blind tz=0 — with zero false-basin acceptances across the sweep. Wider capture range than the alternating pipeline (~5-6°) or a single non-iterated frozen-point shot (~19°). Validated on one real dataset so far.

Tests — tests/test_frozen_point.py, 2 new synthetic tests added to the existing 2:
- test_iterate_frozen_point_until_stable_escapes_blind_seed — confirms the iterative wrapper recovers a known synthetic geometry from a seed a single direct fit cannot solve.
- test_iterate_frozen_point_until_stable_reports_non_convergence_honestly — confirms a seed outside the capture range correctly returns converged=False rather than a false-positive stable basin.

All 4 tests pass.

Notebook 26 (notebooks/26_frozen_point_calibration.ipynb + its _nb26_source.py source, _build.py/README.md registration) — an executed, self-contained walkthrough on a real dataset: loads raw HDF5 frames directly (no external metadata files), runs both new functions, a tilt+distortion-correct interactive Plotly ring overlay, and a distortion-basis comparison (geometry-only vs. isotropic vs. full harmonic).

Testing done

- pytest tests/test_frozen_point.py -v — 4/4 pass.
- Notebook executed end-to-end, zero errors, checked into the PR with real outputs.

Notes for the reviewer

- Both functions mutate their v1_params argument in place (writing the fitted geometry back onto it) — same convention the file already uses.
- The iterative wrapper's capture range is characterized on one real dataset; broader validation across more geometries would strengthen the general recommendation, called out explicitly in the docstring.

### CODE REVIEW

**Review by marinerhemant — COMMENTED** [review]
Thanks — this is a good piece of work, and the structural idea is the right one. Freezing the point cloud removes the extract-refit feedback loop rather than damping it, which is why it tolerates a bad tilt seed where the alternating pipelines don't. **No merge blocker from me.** Everything below is either a comment I'd like changed before merge, or a question about your data.

What I ran: your four tests, and the full `midas_calibrate_v2` suite on `frozen-point-high-tilt-calibration`. Green.

I also like that `test_iterate_frozen_point_until_stable_reports_non_convergence_honestly` exists. A pipeline whose failure mode is "wrong basin, confidently" needs a test that the honest report actually happens, and most PRs don't write it.

---

## The one thing I'd change before merge: the `tx` comment, not the code

Keep the freeze. Rewrite why.

```python
# tx (rotation about the beam axis) is unobservable from a single
# powder image's ring radii alone -- only the azimuthal (eta) labelling
# of a ring changes with tx, never its radius -- so it is a genuine
# gauge freedom for this kind of data, not merely poorly constrained.
```

I believed this too, wrote it down, and then had it refuted — so I'd rather you get the corrected version than inherit mine.

`tx` is an *exact* gauge freedom only in the sub-model where **every azimuthal distortion amplitude is zero**. That is not a detail; it is the only path by which `tx` reaches a radius at all. MIDAS evaluates the distortion at the **lab** azimuth:

```
D(rho, eta) = 1 + iso terms + sum_k a_k * rho^n_k * cos(k*eta' + phi_k),   eta' = 90 - eta
```

`tx` shifts lab `eta` by exactly `tx`, so with any `a_k != 0` it moves `R`. Concretely, with all 15 coefficients pinned at zero the derivative of the pseudo-strain cost w.r.t. `tx` is 1.1e-15 (against 0.87 for `ty` and 19.5 for `tz`) — that is the number that made me write "exact". Switch the distortion on and it isn't: `||J[:,tx]||^2` is a positive-definite quadratic form in the six harmonic amplitudes whose only zero is `a = 0`, and at a real CeO2 calibration `tx` carries roughly **30 microstrain per degree** that cannot be profiled out by `Lsd`, `BC`, `ty` or `tz`.

So `tx` is **weakly observable, and observable only through the distortion** — which is exactly why you should still freeze it, and why the freeze is worth *more* than the original comment claimed:

* with the harmonics **free**, `(tx, phi_k) -> (tx + d, phi_k + k*d)` is an exact gauge orbit. Refining `tx` walks that orbit and corrupts all six phases at once, with no residual signature to warn you.
* with the harmonics **frozen** (what `_bounded_geom_only_spec` does), `tx` is determined — but determined *by the frozen field*, so a field fitted in the wrong frame hands back a confident wrong `tx`.

Suggested replacement:

```python
# tx (rotation about the beam) reaches the ring radii ONLY through the
# azimuthal distortion harmonics: it shifts lab eta by exactly tx, and
# D is evaluated at lab eta. With the harmonics free that makes
# (tx, phi_k) -> (tx + d, phi_k + k*d) an exact gauge orbit -- refining
# tx walks it and corrupts all six phases silently. With them frozen tx
# is determined only by the frozen field, so a field fitted at the wrong
# tx returns a confident wrong tx. Either way a single powder image
# cannot refine it, so it is frozen here regardless of the caller's spec.
```

This also bears on your issue #72 part A, and I'll answer that separately — short version: **seeding `tx` yes, refining it no**, and your own point 3 on #70 is the reason.

---

## `iterate_frozen_point_until_stable` leaves the caller's `v1_params` at iteration 1

Real, and I'll say up front it is **not a blocker** — your docstring already declares the mutation, and I could not find a caller it hurts.

The mechanism: `seed = v1_params` initially, so iteration 1 mutates the caller's object; then `seed = _reseed(seed, fit)` rebinds to a fresh object, so iterations 2..n mutate that instead. On your own fixture: `converged=True` after 9 iterations, returned `.fit` `tz=4.879` (truth 5.000), caller's `v1_params.tz=1.163`, `ty=-0.646` (truth +2.000). Bit-identical to `history[0]` in 130/130 seeds and noise draws, and never equal to `.fit`.

Why it isn't a blocker: no call site in `packages/` reads geometry off a passed-in params object, and all 24 `write_v1_paramstest()` sites launder it through `unpacked_to_v1_params`, which overwrites every geometry field — I fed it the **stale** object as template and still got the converged geometry out.

Still, "mutates in place" plus "the in-place value is not the answer" is a sharp edge. Cheapest fix is one line after the loop:

```python
for k in ("Lsd", "BC_y", "BC_z", "ty", "tz"):
    setattr(v1_params, k, getattr(res.unpacked_v1, k))   # or from `fit`
```

so the mutation at least means what a reader assumes. Your call — documenting it harder is also fine.

---

## Two questions about `point_pick` on your data

Neither is a defect I can demonstrate on your dataset; both are places where I'd want a number before this gets used on a calibrant that matters.

**1. The SNR floor is one scalar per ring, so acceptance is azimuth-dependent.**

```python
ring_vals = img[ring_mask]
baseline  = float(np.median(ring_vals))
mad       = float(np.median(np.abs(ring_vals - baseline)))
```

`baseline` and `noise` are computed over the **whole annulus**, then applied per pixel. Any real azimuthal intensity variation — polarization, detector gain, absorption, a bit of texture in the calibrant — means bright azimuths clear `snr_threshold` and dim ones don't. That biases *which azimuths contribute points*, and this pipeline fits `ty`/`tz`/`BC`, which are precisely the parameters an azimuthally uneven sampling can lean on. It gets worse at large tilt, which is the case the PR is for.

Worth printing points-per-azimuth-octant per ring alongside the existing per-ring counts. If it's flat, this is a non-issue and the print is still useful. Making `baseline`/`noise` local (per azimuthal sector) would fix it if it isn't.

**2. A saturated peak is picked many times.**

`local_max = (img == ndimage.maximum_filter(img, size=footprint_px))` flags **every** pixel of a flat top, because each equals the neighbourhood max. A 5x5 saturated plateau gives 25 flagged pixels (I checked: 25 of 25); sub-pixel refinement then maps them to nearly the same centroid, so one physical peak enters the fit ~25 times at full weight. On a strong low-angle CeO2 ring on a Varex that seems reachable.

Cheap guard: label connected components of `cand_mask` and keep one point per component, or reject any candidate whose patch has zero dynamic range.

---

## Smaller things

* **`autocalibrate_frozen_point` mutates a caller-supplied `spec`.** It sets `spec.parameters["tx"].refined = False` and writes `.init` for every fitted scalar. `iterate_frozen_point_until_stable` builds a fresh spec each pass so it doesn't bite there, but a caller reusing one spec across calls gets silent carry-over. `dataclasses.replace(spec)` or a documented note.

* **`_reseed` aliases mutable state.** `dataclasses.replace` copies field *references*, so every "copy" shares one `.extra` dict and one `.Refine` dict with the caller's object — I verified both (`a.extra is b.extra` -> `True`, and mutating the copy changes the original). Nothing in this PR writes to either during the loop, so it's latent rather than live, but it's the kind of thing that becomes a real bug the first time someone stashes a per-iteration value in `.extra`.

* **`except Exception: pass` in the write-back loop.** In

    ```python
    setattr(v1_params, name, type(cur)(float(val.detach())))
    ```

  `type(cur)` also silently truncates if `cur` is an `int`. A bare swallow around a geometry write-back is worth at least a `LOG.debug`, so a systematic failure to write back isn't invisible.

* **`min_ring_gap_deg: float = 0.0` "for backward compatibility".** This is a brand-new module, so there's no back-compat to keep — and the docstring immediately tells CeO2 users to override it. A default the docs tell you not to use is the wrong default; I'd ship the safe value.

* **Importing `_filter_by_snr` (private) from `single_pv`.** Fine, but it makes `single_pv`'s private surface load-bearing for a second pipeline. Worth promoting to `_common`.

---

## One framing note, not a change request

The four tests all paint the image with the same forward model they then fit, so they're an inverse crime — good evidence about wiring and identifiability, no evidence about accuracy on real data. That's not a criticism; it's the right test to write. I mention it only because the capture-range claim ("roughly a 35 deg range around the true tilt") is measured on that synthetic, and your docstring is already careful to say the real-data validation is one dataset. Keep that sentence — it's doing work.

The `tz ~ 14 deg` CeO2 you validated against is also, as it happens, the most useful dataset either of us has for the open question of whether the distortion field is locked to the panel or to the lab. More on that separately.

**Comment by d-beniwal** [comment]
Thanks for these insights. I will submit a new pull request with these changes incorporated.
